No index signature with a parameter of type 'string' was found on type 'ISpecificationExtension'.

We are trying to update to 4.0.2 and our builds are broken with the following errors after changing my imports as documented.

Changed import to:

```ts
import { MediaTypeObject } from 'openapi3-ts/src/model/openapi31';
```

I get:

```
node_modules/openapi3-ts/src/model/oas-common.ts:21:16 - error TS7053: Element implicitly has an 'any' type because expression of type 'string' can't be used to index type 'ISpecificationExtension'.
  No index signature with a parameter of type 'string' was found on type 'ISpecificationExtension'.

21         return obj[extensionName];
                  ~~~~~~~~~~~~~~~~~~

node_modules/openapi3-ts/src/model/oas-common.ts:31:9 - error TS7053: Element implicitly has an 'any' type because expression of type 'string' can't be used to index type 'ISpecificationExtension'.
  No index signature with a parameter of type 'string' was found on type 'ISpecificationExtension'.

31         obj[extensionName] = extension;
           ~~~~~~~~~~~~~~~~~~

node_modules/openapi3-ts/src/model/specification-extension.ts:28:13 - error TS7053: Element implicitly has an 'any' type because expression of type 'string' can't be used to index type 'SpecificationExtension'.
  No index signature with a parameter of type 'string' was found on type 'SpecificationExtension'.

28         if (this[extensionName]) {
               ~~~~~~~~~~~~~~~~~~~

node_modules/openapi3-ts/src/model/specification-extension.ts:29:20 - error TS7053: Element implicitly has an 'any' type because expression of type 'string' can't be used to index type 'SpecificationExtension'.
  No index signature with a parameter of type 'string' was found on type 'SpecificationExtension'.

29             return this[extensionName];
                      ~~~~~~~~~~~~~~~~~~~

node_modules/openapi3-ts/src/model/specification-extension.ts:39:9 - error TS7053: Element implicitly has an 'any' type because expression of type 'string' can't be used to index type 'SpecificationExtension'.
  No index signature with a parameter of type 'string' was found on type 'SpecificationExtension'.

39         this[extensionName] = payload;
           ~~~~~~~~~~~~~~~~~~~
```

--- a/Changelog.md
+++ b/Changelog.md
@@ -1,5 +1,11 @@
 # Changelog: openapi3-ts
 
+## Version 4.0.3
+
+2023.03.30
+
+- Bug fix. Stricter TS compilation. Fix issue [#105](https://github.com/metadevpro/openapi3-ts/issues/105)
+
 ## Version 4.0.2
 
 2023.03.30

--- a/package.json
+++ b/package.json
@@ -1,6 +1,6 @@
 {
     "name": "openapi3-ts",
-    "version": "4.0.2",
+    "version": "4.0.3",
     "description": "TS Model & utils for OpenAPI 3.x specification.",
     "main": "dist/cjs/index.js",
     "module": "dist/mjs/index.js",
@@ -46,7 +46,7 @@
         "eslint-plugin-prettier": "^4.2.1",
         "prettier": "^2.8.7",
         "rimraf": "^4.4.1",
-        "typescript": "~5.0.2",
+        "typescript": "~5.0.3",
         "vitest": "^0.29.8",
         "vitest-teamcity-reporter": "^0.1.12"
     }

--- a/src/dsl/openapi-builder30.ts
+++ b/src/dsl/openapi-builder30.ts
@@ -103,18 +103,26 @@ export class OpenApiBuilder {
         return this;
     }
     addSchema(name: string, schema: oa.SchemaObject | oa.ReferenceObject): OpenApiBuilder {
+        this.rootDoc.components = this.rootDoc.components || {};
+        this.rootDoc.components.schemas = this.rootDoc.components.schemas || {};
         this.rootDoc.components.schemas[name] = schema;
         return this;
     }
     addResponse(name: string, response: oa.ResponseObject | oa.ReferenceObject): OpenApiBuilder {
+        this.rootDoc.components = this.rootDoc.components || {};
+        this.rootDoc.components.responses = this.rootDoc.components.responses || {};
         this.rootDoc.components.responses[name] = response;
         return this;
     }
     addParameter(name: string, parameter: oa.ParameterObject | oa.ReferenceObject): OpenApiBuilder {
+        this.rootDoc.components = this.rootDoc.components || {};
+        this.rootDoc.components.parameters = this.rootDoc.components.parameters || {};
         this.rootDoc.components.parameters[name] = parameter;
         return this;
     }
     addExample(name: string, example: oa.ExampleObject | oa.ReferenceObject): OpenApiBuilder {
+        this.rootDoc.components = this.rootDoc.components || {};
+        this.rootDoc.components.examples = this.rootDoc.components.examples || {};
         this.rootDoc.components.examples[name] = example;
         return this;
     }
@@ -122,10 +130,14 @@ export class OpenApiBuilder {
         name: string,
         reqBody: oa.RequestBodyObject | oa.ReferenceObject
     ): OpenApiBuilder {
+        this.rootDoc.components = this.rootDoc.components || {};
+        this.rootDoc.components.requestBodies = this.rootDoc.components.requestBodies || {};
         this.rootDoc.components.requestBodies[name] = reqBody;
         return this;
     }
     addHeader(name: string, header: oa.HeaderObject | oa.ReferenceObject): OpenApiBuilder {
+        this.rootDoc.components = this.rootDoc.components || {};
+        this.rootDoc.components.headers = this.rootDoc.components.headers || {};
         this.rootDoc.components.headers[name] = header;
         return this;
     }
@@ -133,22 +145,30 @@ export class OpenApiBuilder {
         name: string,
         secScheme: oa.SecuritySchemeObject | oa.ReferenceObject
     ): OpenApiBuilder {
+        this.rootDoc.components = this.rootDoc.components || {};
+        this.rootDoc.components.securitySchemes = this.rootDoc.components.securitySchemes || {};
         this.rootDoc.components.securitySchemes[name] = secScheme;
         return this;
     }
     addLink(name: string, link: oa.LinkObject | oa.ReferenceObject): OpenApiBuilder {
+        this.rootDoc.components = this.rootDoc.components || {};
+        this.rootDoc.components.links = this.rootDoc.components.links || {};
         this.rootDoc.components.links[name] = link;
         return this;
     }
     addCallback(name: string, callback: oa.CallbackObject | oa.ReferenceObject): OpenApiBuilder {
+        this.rootDoc.components = this.rootDoc.components || {};
+        this.rootDoc.components.callbacks = this.rootDoc.components.callbacks || {};
         this.rootDoc.components.callbacks[name] = callback;
         return this;
     }
     addServer(server: oa.ServerObject): OpenApiBuilder {
+        this.rootDoc.servers = this.rootDoc.servers || [];
         this.rootDoc.servers.push(server);
         return this;
     }
     addTag(tag: oa.TagObject): OpenApiBuilder {
+        this.rootDoc.tags = this.rootDoc.tags || [];
         this.rootDoc.tags.push(tag);
         return this;
     }

--- a/src/dsl/openapi-builder31.ts
+++ b/src/dsl/openapi-builder31.ts
@@ -99,22 +99,31 @@ export class OpenApiBuilder {
         return this;
     }
     addPath(path: string, pathItem: oa.PathItemObject): OpenApiBuilder {
+        this.rootDoc.paths = this.rootDoc.paths || [];
         this.rootDoc.paths[path] = { ...(this.rootDoc.paths[path] || {}), ...pathItem };
         return this;
     }
     addSchema(name: string, schema: oa.SchemaObject | oa.ReferenceObject): OpenApiBuilder {
+        this.rootDoc.components = this.rootDoc.components || {};
+        this.rootDoc.components.schemas = this.rootDoc.components.schemas || {};
         this.rootDoc.components.schemas[name] = schema;
         return this;
     }
     addResponse(name: string, response: oa.ResponseObject | oa.ReferenceObject): OpenApiBuilder {
+        this.rootDoc.components = this.rootDoc.components || {};
+        this.rootDoc.components.responses = this.rootDoc.components.responses || {};
         this.rootDoc.components.responses[name] = response;
         return this;
     }
     addParameter(name: string, parameter: oa.ParameterObject | oa.ReferenceObject): OpenApiBuilder {
+        this.rootDoc.components = this.rootDoc.components || {};
+        this.rootDoc.components.parameters = this.rootDoc.components.parameters || {};
         this.rootDoc.components.parameters[name] = parameter;
         return this;
     }
     addExample(name: string, example: oa.ExampleObject | oa.ReferenceObject): OpenApiBuilder {
+        this.rootDoc.components = this.rootDoc.components || {};
+        this.rootDoc.components.examples = this.rootDoc.components.examples || {};
         this.rootDoc.components.examples[name] = example;
         return this;
     }
@@ -122,10 +131,14 @@ export class OpenApiBuilder {
         name: string,
         reqBody: oa.RequestBodyObject | oa.ReferenceObject
     ): OpenApiBuilder {
+        this.rootDoc.components = this.rootDoc.components || {};
+        this.rootDoc.components.requestBodies = this.rootDoc.components.requestBodies || {};
         this.rootDoc.components.requestBodies[name] = reqBody;
         return this;
     }
     addHeader(name: string, header: oa.HeaderObject | oa.ReferenceObject): OpenApiBuilder {
+        this.rootDoc.components = this.rootDoc.components || {};
+        this.rootDoc.components.headers = this.rootDoc.components.headers || {};
         this.rootDoc.components.headers[name] = header;
         return this;
     }
@@ -133,22 +146,30 @@ export class OpenApiBuilder {
         name: string,
         secScheme: oa.SecuritySchemeObject | oa.ReferenceObject
     ): OpenApiBuilder {
+        this.rootDoc.components = this.rootDoc.components || {};
+        this.rootDoc.components.securitySchemes = this.rootDoc.components.securitySchemes || {};
         this.rootDoc.components.securitySchemes[name] = secScheme;
         return this;
     }
     addLink(name: string, link: oa.LinkObject | oa.ReferenceObject): OpenApiBuilder {
+        this.rootDoc.components = this.rootDoc.components || {};
+        this.rootDoc.components.links = this.rootDoc.components.links || {};
         this.rootDoc.components.links[name] = link;
         return this;
     }
     addCallback(name: string, callback: oa.CallbackObject | oa.ReferenceObject): OpenApiBuilder {
+        this.rootDoc.components = this.rootDoc.components || {};
+        this.rootDoc.components.callbacks = this.rootDoc.components.callbacks || {};
         this.rootDoc.components.callbacks[name] = callback;
         return this;
     }
     addServer(server: oa.ServerObject): OpenApiBuilder {
+        this.rootDoc.servers = this.rootDoc.servers || [];
         this.rootDoc.servers.push(server);
         return this;
     }
     addTag(tag: oa.TagObject): OpenApiBuilder {
+        this.rootDoc.tags = this.rootDoc.tags || [];
         this.rootDoc.tags.push(tag);
         return this;
     }

--- a/src/model/oas-common.ts
+++ b/src/model/oas-common.ts
@@ -1,6 +1,10 @@
 /* eslint-disable @typescript-eslint/no-explicit-any */
 
-import { ISpecificationExtension, SpecificationExtension } from './specification-extension';
+import {
+    IExtensionName,
+    ISpecificationExtension,
+    SpecificationExtension
+} from './specification-extension';
 
 export interface ServerObject extends ISpecificationExtension {
     url: string;
@@ -18,7 +22,7 @@ export function getExtension(obj: ISpecificationExtension | undefined, extension
         return undefined;
     }
     if (SpecificationExtension.isValidExtension(extensionName)) {
-        return obj[extensionName];
+        return obj[extensionName as IExtensionName];
     }
     return undefined;
 }
@@ -28,6 +32,6 @@ export function addExtension(
     extension: any
 ): void {
     if (obj && SpecificationExtension.isValidExtension(extensionName)) {
-        obj[extensionName] = extension;
+        obj[extensionName as IExtensionName] = extension;
     }
 }

--- a/src/model/specification-extension.ts
+++ b/src/model/specification-extension.ts
@@ -25,8 +25,8 @@ export class SpecificationExtension implements ISpecificationExtension {
                 `Invalid specification extension: '${extensionName}'. Extensions must start with prefix 'x-`
             );
         }
-        if (this[extensionName]) {
-            return this[extensionName];
+        if (this[extensionName as IExtensionName]) {
+            return this[extensionName as IExtensionName];
         }
         return null;
     }
@@ -36,7 +36,7 @@ export class SpecificationExtension implements ISpecificationExtension {
                 `Invalid specification extension: '${extensionName}'. Extensions must start with prefix 'x-`
             );
         }
-        this[extensionName] = payload;
+        this[extensionName as IExtensionName] = payload;
     }
     listExtensions(): string[] {
         const res: string[] = [];

--- a/tsconfig.json
+++ b/tsconfig.json
@@ -1,20 +1,25 @@
 {
     "compilerOptions": {
         "declaration": true,
+        "experimentalDecorators": true,
         "target": "ES2020",
+        "outDir": "dist/mjs",
+        "preserveConstEnums": true,
         "module": "ESNext",
         "moduleResolution": "Node16",
-        "experimentalDecorators": true,
-        "noImplicitAny": false,
-        "removeComments": true,
-        "preserveConstEnums": true,
         "noEmitHelpers": true,
+        "noImplicitAny": true,
+        "noImplicitReturns": true,
+        "noUnusedLocals": true,
+        "noUnusedParameters": true,
+        "removeComments": true,
         "skipLibCheck": true,
-        "outDir": "dist/mjs",
+        "strict": true,
+        "sourceMap": true,
         "rootDir": "src",
         "types": ["vitest/globals", "node"]
     },
-    "compileOnSave": true,
     "buildOnSave": true,
+    "compileOnSave": true,
     "exclude": ["**/*.spec.ts", "bin", "dist", "vite.config.ts", "node_modules/**"]
 }

